File: 09Function/closure/output_parameters.rs
/**闭包作为输入参数是可能的，所以返回闭包作为输出参数（output parameter）也应该是可能的。然而返回闭包类型会有问题，因为目前 Rust 只支持返回具体（非泛型）的类型。按照定义，匿名的闭包的类型是未知的，所以只有使用impl Trait才能返回一个闭包。
作为返回值的部分。似乎是个使用策略的好方法啊。也方便一些Job system的代码写法。
*/
fn create_fn() -> impl Fn() {
    let text = "Fn".to_owned();

    move || println!("This is a: {}", text)
}

fn create_fnmut() -> impl FnMut() {
    let text = "FnMut".to_owned();

    move || println!("This is a: {}", text)
}

fn create_fnonce() -> impl FnOnce() {
    let text = "FnOnce".to_owned();
    
    move || println!("This is a: {}", text)
}

fn main() {
    let fn_plain = create_fn();
    let mut fn_mut = create_fnmut();
    let fn_once = create_fnonce();

    fn_plain();
    fn_mut();
    fn_once();
}

File: 09Function/closure/output_parameters.cs
using System;

class Program
{
    static Action CreateFn()
    {
        string text = "Fn";
        return () => Console.WriteLine($"This is a: {text}");
    }

    static Action CreateFnMut()
    {
        string text = "FnMut";
        return () => Console.WriteLine($"This is a: {text}");
    }

    static Action CreateFnOnce()
    {
        string text = "FnOnce";
        bool called = false;
        return () =>
        {
            if (!called)
            {
                Console.WriteLine($"This is a: {text}");
                called = true;
            }
            else
            {
                throw new InvalidOperationException("This action can only be called once.");
            }
        };
    }

    static void Main(string[] args)
    {
        Console.WriteLine("This is C# Language");
        var fnPlain = CreateFn();
        var fnMut = CreateFnMut();
        var fnOnce = CreateFnOnce();

        fnPlain(); // Can be invoked multiple times.
        fnMut();   // Can also be invoked multiple times.
        fnOnce();  // Designed to be invoked only once.

        // Uncommenting the following line will throw an exception because fnOnce() can only be called once.
        //fnOnce();
    }
}